sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/TableHeaderReferenceCheck.java

```diff
@@ -0,0 +1,398 @@
+          currentRow.set(columnIndex, column);
+        } else {
+          currentRow.add(column);
+        }
+        colspan--;
+      }
+    }
+
+    private void newColumnWithRowSpan(Table.Column column) {
+      int rowspan = getRowSpan(column.node());
+      int rowStart = rows.indexOf(currentRow);
+      int rowEnd = rowStart + rowspan;
+      int columnIndex = rows.get(rowStart).indexOfVacantColumn();
+      if (columnIndex == -1) {
+        columnIndex = rows.get(rowStart).size();
+      }
+      for (int row = rowStart; row < rowEnd; ++row) {
+        if (row == rows.size()) {
+          rows.add(new RowBuilder());
+        }
+        if (columnIndex < rows.get(row).size()) {
+          rows.get(row).set(columnIndex, column);
+        } else {
+          for (int i = rows.get(row).size(); i < columnIndex; ++i) {
+            rows.get(row).add(null);
+          }
+          rows.get(row).add(column);
+        }
+      }
+    }
+
+    private void newColumnWithRowColSpan(Table.Column column) {
+      int rowspan = getRowSpan(column.node());
+      int rowStart = rows.indexOf(currentRow);
+      int rowEnd = rowStart + rowspan;
+      int colspan = getColSpan(column.node());
+      int columnStart = rows.get(rowStart).indexOfVacantColumn();
+      if (columnStart == -1) {
+        columnStart = rows.get(rowStart).size();
+      }
+      int columnEnd = columnStart + colspan;
+      for (int row = rowStart; row < rowEnd; ++row) {
+        if (row == rows.size()) {
+          rows.add(new RowBuilder());
+        }
+        for (int col = columnStart; col < columnEnd; ++col) {
+          if (col < rows.get(row).size()) {
+            rows.get(row).set(col, column);
+          } else {
+            for (int i = rows.get(row).size(); i < col; ++i) {
+              rows.get(row).add(null);
+            }
+            rows.get(row).add(column);
+          }
+        }
+      }
+    }
+
+    public Table build() {
+      return new Table(rows.stream().map(RowBuilder::build).collect(Collectors.toList()));
+    }
+  }
+
+  @Override
+  public void startDocument(List<Node> nodes) {
+    stack.clear();
+  }
+
+  @Override
+  public void startElement(TagNode node) {
+    if (isTable(node)) {
+      stack.push(new TableBuilder());
+    } else if (!stack.isEmpty()) {
+      if (isTableRow(node)) {
+        stack.peek().newRow();
+      } else if (isTableColumn(node)) {
+        stack.peek().newColumn(new Table.Column(node));
+      } else if (isTableHeader(node)) {
+        stack.peek().newColumn(new Table.Header(node));
+      }
+    }
+  }
+
+  @Override
+  public void endElement(TagNode node) {
+    if (isTable(node) && !stack.isEmpty()) {
+      raiseViolationOnInvalidReference(stack.pop().build());
+    }
+  }
+
+  private void raiseViolationOnInvalidReference(Table table) {
+    Map<TagNode, List<String>> referenceableHeaders = findReferenceableHeadersPerColumnNode(table);
+    for (int row = 0; row < table.rows().size(); ++row) {
```

Comment:
This loop pattern
```
 for (int row = 0; row < table.rows().size(); ++row) {
      for (int col = 0; col < table.rows().get(row).size(); ++col) {
        Table.Column column = table.rows().get(row).get(col);
```
is repeated in four methods. Maybe we can have a method in `Table` to provide this iteration